src/samVisualize/rendering/forceDirectedGraph.ts

```diff
@@ -0,0 +1,703 @@
+            .style('font-weight', 'bold')
+            .text('All resources')
+
+        return buttonDiv
+    }
+
+    // Render a GraphObject
+    private update(newGraphObject: GraphObject | undefined) {
+        // Never occurs. Possibly undefined args are used in calls within a listener declaration,
+        // but those listener events are only fired with verified GraphObjects
+        if (newGraphObject === undefined) {
+            return
+        }
+
+        // Clear the graph
+        this.g.selectAll('g').remove()
+
+        this.links = this.constructLinks(this.g, newGraphObject.links)
+        this.nodes = this.constructNodes(this.g, newGraphObject.nodes, this.iconPaths, this.vscodeApi)
+
+        // Implement drag behavior
+        this.nodes.call(
+            d3
+                .drag<SVGGElement, NodeDatum>()
+                .on('start', this.dragStarted)
+                .on('drag', this.dragged)
+                .on('end', this.dragEnded)
+        )
+        // Attach nodes & links to simulation, upon which specified forces are applied
+        this.simulation.nodes(newGraphObject.nodes).on('tick', () => this.ticked())
+        this.simulation.force<d3.ForceLink<NodeDatum, LinkDatum>>('link')?.links(newGraphObject.links)
+
+        // Give energy to a new render
+        this.simulation.alphaTarget(RenderConstants.reheatAlphaTarget).restart()
+    }
+    private dragStarted = (event: d3.D3DragEvent<SVGGElement, NodeDatum, unknown>, d: NodeDatum): void => {
+        if (!event.active) {
+            this.simulation.alphaTarget(RenderConstants.reheatAlphaTarget).restart()
+        }
+        d.fx = d.x
+        d.fy = d.y
+    }
+
+    private dragged = (event: d3.D3DragEvent<SVGGElement, NodeDatum, unknown>, d: NodeDatum): void => {
+        this.simulation.alpha(RenderConstants.duringDragAlpha)
+        d.fx = event.x
+        d.fy = event.y
+    }
+
+    private dragEnded = (event: d3.D3DragEvent<SVGGElement, NodeDatum, unknown>, d: NodeDatum): void => {
+        delete d.fx
+        delete d.fy
+    }
+
+    /**
+     * Updates link paths and node positions every tick.
+     */
+    public ticked(): void {
+        this.links!.selectAll<SVGGElement, LinkDatum>('path').attr('d', this.calculateLinkPath)
+        // // Create a path for the tooltip assistance link, which is just a thicker invisible line
+        // // drawn over existing links (same 'd' value) to make it easier to get a tooltip while hovering over a link.
+        this.nodes!.attr('transform', (d: NodeDatum) => `translate(${d.x}, ${d.y})`)
+        // Once we hit the target after a reheat, begin decaying to long term alpha target
+        if (Math.abs(this.simulation.alpha() - RenderConstants.reheatAlphaTarget) < 0.001) {
+            this.simulation.alphaTarget(RenderConstants.alphaTarget).restart()
+        }
+    }
+
+    /**
+     * Scales down a vector to account for the radius of the target node.
```

Comment:
Links have different widths?

Reply:
Well that function is so that links point to the edge of the target node, and not the centre.